Safari crashes when showing the downloads menu for a download with no MIME type

hortont424 opened the pull request:
#### d94ee98c76194e8e618a849534b8401c984f3c7f
<pre>
Safari crashes when showing the downloads menu for a download with no MIME type
<a href="https://bugs.webkit.org/show_bug.cgi?id=259622">https://bugs.webkit.org/show_bug.cgi?id=259622</a>
rdar://113058721

Reviewed by Wenson Hsieh.

* Source/WebCore/loader/mac/LoaderNSURLExtras.mm:
Add a null check.

* Tools/TestWebKitAPI/Tests/WebKitLegacy/mac/URLExtras.mm: Renamed from Tools/TestWebKitAPI/Tests/WebKitLegacy/mac/URLCanonicalization.mm.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
Rename URLCanonicalization to URLExtras and add a (very simple) positive and
negative test for _webkit_suggestedFilenameWithMIMEType, which was what
was triggering this crash.

Canonical link: <a href="https://commits.webkit.org/266414@main">https://commits.webkit.org/266414@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3462a94965086644b889c4e7fcc8a3895edde8a3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13775 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15511 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13083 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13856 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16597 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14170 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15763 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13942 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14562 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11667 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16213 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11850 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12425 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19464 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12926 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12590 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15808 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13119 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10997 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12388 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3348 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16721 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12961 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitLegacy/mac/URLExtras.mm:
```diff
@@ -25,6 +25,7 @@
 #import "config.h"
 
+#import "PlatformUtilities.h"
 #import "Test.h"
```
why is this file in WebKitLegacy?

whsieh replied:
That's probably because the code that it exercises (`-_webkit_suggestedFilenameWithMIMEType`) lives in WebKitLegacy.

hortont424 replied:
Quite!

hortont424 committed:
Safari crashes when showing the downloads menu for a download with no MIME type
https://bugs.webkit.org/show_bug.cgi?id=259622
rdar://113058721

Reviewed by Wenson Hsieh.

* Source/WebCore/loader/mac/LoaderNSURLExtras.mm:
Add a null check.

* Tools/TestWebKitAPI/Tests/WebKitLegacy/mac/URLExtras.mm: Renamed from Tools/TestWebKitAPI/Tests/WebKitLegacy/mac/URLCanonicalization.mm.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
Rename URLCanonicalization to URLExtras and add a (very simple) positive and
negative test for _webkit_suggestedFilenameWithMIMEType, which was what
was triggering this crash.

Canonical link: https://commits.webkit.org/266414@main
